spec/index.bs

```diff
@@ -587,107 +587,114 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
 
 ### Resource Usages ### {#programming-model-resource-usages}
 
-Buffers and textures can be used by the GPU in multiple ways,
-which can be split into two groups:
-
-: <dfn dfn>Read-only usage</dfn>s
-:: Usages like
-    {{GPUBufferUsage/VERTEX|GPUBufferUsage.VERTEX}} or
-    {{GPUTextureUsage/SAMPLED|GPUTextureUsage.SAMPLED}}
-    don't change the contents of a resource.
-
-: <dfn dfn>Mutating usage</dfn>s
-:: Usages like {{GPUBufferUsage/STORAGE|GPUBufferUsage.STORAGE}}
-    do change the contents of a resource.
-
-Issue(gpuweb/gpuweb#296): Consider merging all read-only usages.
+A [=physical resource=] can be used on GPU in one of the following <dfn dfn>internal usage</dfn>s:
+<dl dfn-type=dfn dfn-for="internal usage">
+    : <dfn>input</dfn>
+    :: Buffer with input data for draw or dispatch calls. Preserves the contents.
+        Allowed by {{GPUBufferUsage/INDEX}}, {{GPUBufferUsage/VERTEX}}, or {{GPUBufferUsage/INDIRECT}}.
+    : <dfn>constant</dfn>
+    ::  Resource bindings that are constant from the shader point of view. Preserves the contents.
+        Allowed by {{GPUBufferUsage/UNIFORM}} or {{GPUTextureUsage/SAMPLED}}.
+    : <dfn>storage</dfn>
+    ::  Writable storage resource binding.
+        Allowed by {{GPUBufferUsage/STORAGE}} or {{GPUTextureUsage/STORAGE}}.
+    : <dfn>storage-read</dfn>
+    ::  Read-only storage resource bindings. Preserves the contents.
+        Allowed by {{GPUBufferUsage/STORAGE}} or {{GPUTextureUsage/STORAGE}}.
+    : <dfn>attachment</dfn>
+    :: Texture used as an output attachment in a render pass.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+    : <dfn>attachment-read</dfn>
+    ::  Texture used as a read-only attachment in a render pass. Preserves the contents.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+</dl>
 
 Textures may consist of separate [=mipmap levels=] and [=array layers=],
 which can be used differently at any given time.
-Each such <dfn dfn>subresource</dfn> is uniquely identified by a
+Each such <dfn dfn>texture subresource</dfn> is uniquely identified by a
 [=texture=], [=mipmap level=], and
 (for {{GPUTextureDimension/2d}} textures only) [=array layer=].
 
-The **main usage rule** is that any [=subresource=]
-at any given time can only be in either:
-  - a combination of [=read-only usage=]s
-  - a single [=mutating usage=]
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
+
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage set</dfn>s, where each usage is compatible with the other:
+  - {[=internal usage/input=] | [=internal usage/constant=] | [=internal usage/storage-read=]} - in any combination
+  - {[=internal usage/storage=] | [=internal usage/storage-read=]} - in any combination
```

Comment:
For buffers it makes sense, but do we really want to have `writeonly-storage` and `readonly-storage` usable in the same scope for textures?

Reply:
We talked about this a bit recently with Kai. It would be weird to forbid this, because the restriction is very arbitrary.
Somebody may hit this and wonder why we put the restriction in.

Reply:
My reasoning is that `readonly-storage-buffer` should always be strictly "better" (less restrictive) than `storage-buffer`. That is, if a storage buffer is only used as readonly, then changing the binding type to be readonly shouldn't break anything. But if it's not compatible with `storage-buffer` then it could break usage compatibility with some other `storage-buffer` usage in the scope.